bbs/bbs.go

```diff
@@ -0,0 +1,892 @@
+	proofLenFloor := 2*octetPointLen + 3*octetScalarLen
+	if len(data) < proofLenFloor {
+		return Proof{}, fmt.Errorf("bbs: malformed proof")
+	}
+
+	// // Points (i.e., (Abar, Bbar) in ProofGen) de-serialization.
+	// 3.  index = 0
+	// 4.  for i in range(0, 1):
+	// 5.      end_index = index + octet_point_length - 1
+	// 6.      A_i = octets_to_point_g1(proof_octets[index..end_index])
+	// 7.      if A_i is INVALID or Identity_G1, return INVALID
+	// 8.      index += octet_point_length
+	// index := 0
+
+	index := 0
+
+	// Abar
+	octets := data[index : index+octetPointLen]
+	Abar := &pairing.G1{}
+	Abar.SetBytes(octets)
+	index += octetPointLen
+
+	// Bbar
+	octets = data[index : index+octetPointLen]
+	Bbar := &pairing.G1{}
+	Bbar.SetBytes(octets)
+	index += octetPointLen
+
+	// Scalars (i.e., (r2^, r3^, m^_j1, ..., m^_jU, c) in
+	// r2h and r3h
+	r2h := &pairing.Scalar{}
+	r2h.SetBytes(data[index : index+octetScalarLen])
+	index += octetScalarLen
+	r3h := &pairing.Scalar{}
+	r3h.SetBytes(data[index : index+octetScalarLen])
+	index += octetScalarLen
+
+	i := 0
+	scalars := make([]*pairing.Scalar, 0)
+	for {
+		if index < len(data) {
+			// XXX(caw): need to check if there is enough data
+			scalars = append(scalars, &pairing.Scalar{})
+			scalars[i].SetBytes(data[index : index+octetScalarLen])
+			index += octetScalarLen
+			i += 1
+		} else {
+			if index != len(data) {
+				return Proof{}, fmt.Errorf("bbs: malformed proof")
+			}
+			if len(scalars) < 3 {
+				return Proof{}, fmt.Errorf("bbs: malformed proof")
+			}
+			return Proof{
+				Abar:        Abar,
+				Bbar:        Bbar,
+				r2h:         r2h,
+				r3h:         r3h,
+				commitments: scalars[0 : len(scalars)-1],
+				c:           scalars[len(scalars)-1],
+			}, nil
+		}
+	}
+
+	// // ProofGen) de-serialization.
+	// 9.  j = 0
+	// 10. while index < length(proof_octets):
+	// 11.     end_index = index + octet_scalar_length - 1
+	// 12.     s_j = OS2IP(proof_octets[index..end_index])
+	// 13.     if s_j = 0 or if s_j >= r, return INVALID
+	// 14.     index += octet_scalar_length
+	// 15.     j += 1
+
+	// 16. if index != length(proof_octets), return INVALID
+	// 17. msg_commitments = ()
+	// 18. If j > 3, set msg_commitments = (s_2, ..., s_(j-2))
+	// 19. return (A_0, A_1, s_0, s_1, msg_commitments, s_(j-1))
+
+	// return Proof{}, nil
+}
+
+type SecretKey struct {
+	sk *pairing.Scalar
+}
+
+type PublicKey []byte
+
+func (s SecretKey) Public() PublicKey {
+	W := pairing.G2Generator()
+	W.ScalarMult(s.sk, W)
+
+	// https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bbs-signatures-03#section-6.2.2-3
+	return W.BytesCompressed()
+}
+
+func (s SecretKey) Encode() []byte {
+	enc, err := s.sk.MarshalBinary()
+	if err != nil {
+		panic(err)
+	}
+	return enc
+}
+
+// Key generation
+// https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bbs-signatures-03#name-key-generation-operations
+func KeyGen(ikm []byte, keyInfo, keyDst []byte) (SecretKey, error) {
```

Comment:
Given this first argument should be the output from crypto/rand.Reader, should there have an internal form for testing and then this function reads from the reader?

Reply:
Yeah, accepting a crypto/rand.Reader instead of raw bytes is better here. I passed the raw seed in for now to expedite checking against the test vectors.